### PR TITLE
Tell pegasus about output map files in the dax

### PR DESCRIPTION
**Pull request by duncan-brown**
Pegasus 4.7.3 and above run ``pegasus-plan`` for sub-daxes under the kickstart script for better job management. This means that pegasus needs to be told about output map files in the DAX so that it can manage them. This is implemented by this patch. This should also fix @tdent's observation that data-reuse never seems to work if the reuse cache contains a ``.dax`` or ``.map`` file.

This is a fix for https://github.com/ligo-cbc/pycbc/issues/1448

For top-level workflows, the output map is tagged as a file needed by any sub-dax job and *also* added to the list of PFNs (since it is created by the workflow generation script not within the DAX):
```
<adag xmlns="http://pegasus.isi.edu/schema/DAX" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://pegasus.isi.edu/schema/DAX http://pegasus.isi.edu/schema/dax-3.6.xsd" version="3.6" name="o2-analysis-3-vdf_c37d16-p3">
        <file name="o2-analysis-3-vdf_c37d16-p3-finalization.dax">
                <pfn url="/home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/o2-analysis-3-vdf_c37d16-p3-finalization.dax" site="local"/>
        </file>
        <file name="o2-analysis-3-vdf_c37d16-p3-main.map">
                <pfn url="/home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/o2-analysis-3-vdf_c37d16-p3-main.map" site="local"/>
        </file>
        <file name="o2-analysis-3-vdf_c37d16-p3-finalization.map">
                <pfn url="/home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/o2-analysis-3-vdf_c37d16-p3-finalization.map" site="local"/>
        </file>
        <file name="o2-analysis-3-vdf_c37d16-p3-main.dax">
                <pfn url="/home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/o2-analysis-3-vdf_c37d16-p3-main.dax" site="local"/>
        </file>
        <dax id="ID0000001" file="o2-analysis-3-vdf_c37d16-p3-main.dax">
                <argument>-Dpegasus.dir.storage.mapper.replica.file=o2-analysis-3-vdf_c37d16-p3-main.map -Dpegasus.dir.storage.mapper.replica=File --output-site local --cleanup inplace --cluster label,horizontal -vvv --cache /home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/_reuse.cache</argument>
                <uses name="o2-analysis-3-vdf_c37d16-p3-main.map" link="input"/>
        </dax>
        <dax id="ID0000002" file="o2-analysis-3-vdf_c37d16-p3-finalization.dax">
                <argument>-Dpegasus.dir.storage.mapper.replica.file=o2-analysis-3-vdf_c37d16-p3-finalization.map -Dpegasus.dir.storage.mapper.replica=File --output-site local --cleanup inplace --cluster label,horizontal -vvv --cache /home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/_reuse.cache</argument>
                <uses name="o2-analysis-3-vdf_c37d16-p3-finalization.map" link="input"/>
        </dax>
        <child ref="ID0000002">
                <parent ref="ID0000001"/>
        </child>
</adag>
```

For sub-DAXes that are generated within the DAX, the output map is just added to the sub-dax planning job and not included as a PFN, since Pegasus knows which job made it and from whence it came:
```
        <job id="ID0008744" name="foreground_minifollowup-H1L1_ID92">
                <argument>--config-files <file name="FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULLforeground_minifollowup.ini"/> --bank-file <file name="H1L1-BANK2HDF-1166486417-2620801.hdf"/> --statmap-file <file name="H1L1-STATMAP_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-1166486417-2620801.hdf"/> --inspiral-segments <file name="H1L1-INSP_SEGMENTS-1166486417-2620801.xml"/> --inspiral-data-read-name DATA_ANALYSED --inspiral-data-analyzed-name TRIGGERS_GENERATED --output-file <file name="H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax"/> --output-map <file name="H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax.map"/> --workflow-name H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax --output-dir /home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/results/7._open_box_result/7.02_loudest_event_followup --single-detector-triggers H1:H1-HDF_TRIGGER_MERGE_FULL_DATA-1166486417-2620801.hdf L1:L1-HDF_TRIGGER_MERGE_FULL_DATA-1166486417-2620801.hdf </argument>
                <profile namespace="dagman" key="category">foreground_minifollowup</profile>
                <profile namespace="hints" key="execution.site">local</profile>
                <profile namespace="condor" key="getenv">True</profile>
                <uses name="H1L1-INSP_SEGMENTS-1166486417-2620801.xml" link="input" register="false" transfer="true"/>
                <uses name="H1L1-BANK2HDF-1166486417-2620801.hdf" link="input" register="false" transfer="true"/>
                <uses name="H1L1-STATMAP_FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULL-1166486417-2620801.hdf" link="input" register="false" transfer="true"/>
                <uses name="FULL_DATA_FULL_CUMULATIVE_CAT_12H_FULL_DATA_FULLforeground_minifollowup.ini" link="input" register="false" transfer="true"/>
                <uses name="L1-HDF_TRIGGER_MERGE_FULL_DATA-1166486417-2620801.hdf" link="input" register="false" transfer="true"/>
                <uses name="H1-HDF_TRIGGER_MERGE_FULL_DATA-1166486417-2620801.hdf" link="input" register="false" transfer="true"/>
                <uses name="H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax" link="output" register="true" transfer="true"/>
                <uses name="H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax.map" link="output" register="true" transfer="true"/>
        </job>
        <dax id="ID0008745" file="H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax">
                <argument>--basename H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801 -Dpegasus.dir.storage.mapper.replica.file=H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax.map -Dpegasus.dir.storage.mapper.replica=File --output-site local --cleanup inplace --cluster label,horizontal -vvv --cache /home/dbrown/projects/aligo/o2/analysis-3/o2-analysis-3-vdf_c37d16-p3/output/_reuse.cache</argument>
                <uses name="H1L1-FOREGROUND_MINIFOLLOWUP_FULL_DATA_FULL_CUMULATIVE_CAT_12H-1166486417-2620801.dax.map" link="input"/>
        </dax>
```

### CODE REVIEW

**Comment by duncan-brown** [comment]
@spxiwh hold on, I messed something up...

**Comment by spxiwh** [comment]
I'm not sure that this will fix the issue that we have to avoid using .dax and .map files when reusing data products! If you think this will, can you test that?

**Comment by duncan-brown** [comment]
@spxiwh if the dax and map files exist, the correct behavior should be to just run the minifollowup workflow, right? It should just skip the sub-workflow creation step?

**Comment by spxiwh** [comment]
Well if I reuse the .map file, it'll put all the output files in the *original* run directory ... But there was some issue with reusing .dax files, which I'm not sure I remember. Would be good to test what happens if trying this before suggesting this be done ...

A potential worry: Before reusing these files caused a failure. Now it might run, but *might* do something weird. This could be dangerous if someone does reuse such files and not know what it will do.

**Comment by duncan-brown** [comment]
@spxiwh test run of regular re-use at

https://sugwg-osg.phy.syr.edu/pegasus/u/dbrown/r/417/w?wf_uuid=98af8976-3829-43a3-b0a6-c27076c2383e

https://sugwg-jobs.phy.syr.edu/~dbrown/O2/o2-analysis-3-vdf_c37d16-p3/

File: pycbc/workflow/core.py [hunk 1]
```diff
@@ -589,7 +590,7 @@ def output_map(self):
         
     @property
     def staging_site(self):  
-        if self.in_workflow != False:
+        if self.in_workflow is not False:
```
Comment (spxiwh) [inline pycbc/workflow/core.py]: Why not just
`if self.in_workflow`
??

File: pycbc/workflow/core.py [hunk 2]
```diff
@@ -580,7 +581,7 @@ def __init__(self, args, name):
  
     @property
     def output_map(self):  
-        if self.in_workflow != False:
+        if self.in_workflow is not False:
```
Comment (spxiwh) [inline pycbc/workflow/core.py]: Why not just:
```
if self.in_workflow
```
??

**Comment by duncan-brown** [comment]
@spxiwh ah yes, reusing map files will write the output in the wrong place, unless the map file is change... And since the minfollowup workflow creation job writes both the map and the dax, then if the map is missing it will be re-run anyway.

OK, this doesn't fix the dax/map re-use problem, but it does fix the 4.7.3 issue. Ready for review.

File: pycbc/workflow/core.py [hunk 3]
```diff
@@ -638,19 +639,32 @@ def execute_node(self, node, verbatim_exe = False):
...
+        # access to a file in os.getcwd() but this code is fragile.
+        job.addArguments('--cache %s' % os.path.join(os.getcwd(), '_reuse.cache'))
+
         if staging_site:
             job.addArguments('--staging-site %s' % staging_site)
             
     def save(self, filename=None, output_map=None, staging_site=None):
         if output_map is None:
             output_map = self.output_map
+            if self.in_workflow is not False:
```
Comment (spxiwh) [inline pycbc/workflow/core.py]:
```
if not self.in_workflow is not False:
    pass
else:
```

File: pycbc/workflow/core.py (hunk 1 above)
Comment (duncan-brown) [inline pycbc/workflow/core.py]: Same here.

File: pycbc/workflow/core.py (hunk 3 above)
Comment (duncan-brown) [inline pycbc/workflow/core.py]: That would be incorrect, I think. If you are in a workflow then them map for that workflow must be added to the parent as it is the parent that plans.

Comment (duncan-brown) [inline pycbc/workflow/core.py]: Oh, I missed the triple negative... Too British for me...

Comment (ahnitz) [inline pycbc/workflow/core.py]: @duncan-brown I think I'm confused by these lines. How does the parent workflow know this is the map file to use when planning the sub-workflow? I see no connection between the two.

**Comment by duncan-brown** [comment]
@ahnitz I saw a comment over email that I don't see in the page. The connection comes from the way that the sub-dax is set up. The DAX fragments above demonstrate that this works.

File: pycbc/workflow/core.py (hunk 3 above)
Comment (ahnitz) [inline pycbc/workflow/core.py]: @duncan-brown Ok, sorry, I figured out what you are doing. Can you move these lines though from save to set_job_properties. I'd like to keep the Pegasus hacks in one place and they all relate to map file handling (input and output).

**Comment by duncan-brown** [comment]
@spxiwh another test run with mass bins (which should complete more quickly than denty bins)

https://sugwg-osg.phy.syr.edu/pegasus/u/dbrown/r/420/w?wf_uuid=9a970de6-5a3d-48d8-936b-2221a86c0aea

File: pycbc/workflow/core.py (hunk 3 above)
Comment (duncan-brown) [inline pycbc/workflow/core.py]: @ahnitz I tried that, but it does not work. ``job`` doesn't have the right attributes and if I do it that way I would have to change the arguments of ``set_job_properties``.  The changes would be much more extensive (every use of ``set_job_properties`` would have to be changed). The way we do this is something of a hack anyway.

**Comment by duncan-brown** [comment]
@spxiwh Re: dax/map, I added some "don't do it" instructions to the docs.

**Comment by duncan-brown** [comment]
@spxiwh @ahnitz test run complete:

https://sugwg-osg.phy.syr.edu/pegasus/u/dbrown/r/420/w?wf_uuid=9a970de6-5a3d-48d8-936b-2221a86c0aea

https://sugwg-jobs.phy.syr.edu/~dbrown/O2/o2-analysis-3-o1-bin-stat-p4/

Please can one of you approve so I can merge? Then @vahi can release Pegasus 4.7.4.

**Pull request merged**